hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java

```diff
@@ -95,10 +95,19 @@ public SparkRDDWriteClient(HoodieEngineContext context, HoodieWriteConfig writeC
   public SparkRDDWriteClient(HoodieEngineContext context, HoodieWriteConfig writeConfig,
                              Option<EmbeddedTimelineService> timelineService) {
     super(context, writeConfig, timelineService);
+    bootstrapMetadataTable();
+  }
+
+  private void bootstrapMetadataTable() {
     if (config.isMetadataTableEnabled()) {
-      // If the metadata table does not exist, it should be bootstrapped here
-      // TODO: Check if we can remove this requirement - auto bootstrap on commit
-      SparkHoodieBackedTableMetadataWriter.create(context.getHadoopConf().get(), config, context);
+      // Defer bootstrap if upgrade / downgrade is pending
+      HoodieTableMetaClient metaClient = createMetaClient(true);
+      UpgradeDowngrade upgradeDowngrade = new UpgradeDowngrade(
+          metaClient, config, context, SparkUpgradeDowngradeHelper.getInstance());
+      if (!upgradeDowngrade.needsUpgradeOrDowngrade(HoodieTableVersion.current())) {
```

Comment:
@prashantwason : can you help me understand something. if upgrade is required, we defer the bootstrapping. So, when exactly the upgrade happens? in a subsequent commit? so IIUC, on first commit where upgrade happens, metadata won't be bootstrapped and with 2 -> 3, we will delete the metadata table. and on 2nd commit, will trigger the actual bootstrap of metadata table.

Reply:
Yes, the bootstrap will happen the next time the SparkRDDWRiteCLient is created (probably in the next clean). 

Currently this is what happens (assuming an existing Table with version 2 and using 0.10 code which has version 3):
1. SparkRDDWriteClient constructor - finds no table so bootstrap it (wasted bootstrap)
2. SparkRDDWriteClient.insert() - runs upgrade code in getTableAndXXX() and there the metadata table is deleted.

Next run:
1. SparkRDDWriteClient constructor - finds no table so bootstrap it (second bootstrap)


For file listing this wasted bootstrap is kinda ok but other indexes if enabled together (e.g. record-level-index enabled with metadata table), then this is a lot of wasted time.

Reply:
gotcha. wondering if we can move the instantiation of metadata table 
```
SparkHoodieBackedTableMetadataWriter.create(context.getHadoopConf().get(), config, context);
```
just after upgrade downgrade step within getTableAndInitCtx. 
So, that we may not miss to bootstrap the table even if upgrade is required. 

Here is what I am thinking. I would assume typically users might stop all processes and then do an upgrade for next hudi version. Once first commit goes through, users might want to get started w/ multi writers. so, would prefer to do the bootstrap on the first write operation where upgrade is triggered. 
Open to hear your thoughts.

Reply:
I tried that step but it did not work because:
1. When getTableAndInitCtx is called an action is already started on the table
2. Metadata bootstrap does not happen because it detects an in-progress action 

Bootstrap in the constructor is surely not ideal. 

Possible ways:
1. Make the bootstrap aware of the "current" operation so it can neglect it. Then we can bootstrap right after upgrade/downgrade step (as you suggested).
2. Bootstrap automatically before write-to-metadata step. 

I prefer #1 too as it is cleaner and metadata table will be available before any actions on the dataset start.

Reply:
thanks for the quick turn around. yes. Let's go with 1.

Reply:
I have implemented #1.

Reply:
thanks!